Runtime/Scripts/MobiledgeXSettings.cs

```diff
@@ -1,14 +1,167 @@
 ﻿
 using UnityEngine;
-
+using DistributedMatchEngine;
+using System.Collections.Generic;
+using System.Linq;
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
 namespace MobiledgeX
 {
-	 public class MobiledgeXSettings: ScriptableObject
-	{
-	    public string orgName;
-	    public string appName;
-	    public string appVers;
-	    public int TCP_Port;
-	    public int UDP_Port;
-	}
+    // Complier Directive if UNITY_EDITOR is used extensively since MobiledgeXSettings Object is used in production 
+    // Without this complier directive , Unity wont build the app
+#if UNITY_EDITOR
+
+    //using PropertyDrawer interface (used for creating custom views of fields in the inspector)
+    //ReadOnly Attribute makes mapped ports' fields immutable in the inspector
+    public class ReadOnlyAttribute : PropertyAttribute{  }
+
+    [CustomPropertyDrawer(typeof(ReadOnlyAttribute))]
+    public class ReadOnlyDrawer : PropertyDrawer
+    {
+        public override float GetPropertyHeight(SerializedProperty property,
+                                                GUIContent label)
+        {
+            return EditorGUI.GetPropertyHeight(property, label, true);
+        }
+
+        public override void OnGUI(Rect position,
+                                   SerializedProperty property,
+                                   GUIContent label)
+        {
+            GUI.enabled = false;
+            EditorGUI.PropertyField(position, property, label, true);
+            GUI.enabled = true;
+        }
+    }
+#endif
+    // MatchingEngine AppPort is not Serializable
+    // Protocol enum is used to override LProto easier to read
+    [System.Serializable]
+    public class Port
+    {
+        /// <summary>
+        /// Tag to label certain tags, you can retrieve ports by tag 
+        /// </summary>
+        public string tag;
+
+#if UNITY_EDITOR
+        [ReadOnly]
+#endif
+        public Protocol protocol;
+#if UNITY_EDITOR
+        [ReadOnly]
+#endif
+        public string fqdn_prefix;
+#if UNITY_EDITOR
+        [ReadOnly]
+#endif
+        public string path_prefix;
+#if UNITY_EDITOR
+        [ReadOnly]
+#endif
+        public int internal_port;
+#if UNITY_EDITOR
```

Comment:
What are all these UNITY_EDITOR compiler checks for?

Reply:
![Uploading Screen Shot 2020-05-15 at 2.01.36 PM.png…]()

Reply:
ReadOnly Attribures for making mapped ports immutable in UnityEditor, the Same Object (MobiledgeXSettings is used during Runtime on builds) so it cant have UnityEditor functions nor attributes.